lib/opam.ml

```diff
@@ -313,23 +313,42 @@ module Url = struct
         else OS.File.must_exist distrib_file >>= fun _ -> assert false
 end
 
-let opam_version () =
-  let v =
-    OS.Cmd.run_out Cmd.(cmd % "--version") |> OS.Cmd.out_string ~trim:true
-  in
-  let of_str = function
-    | "1.2.2" -> `v1_2_2
-    | s -> (
-        match String.cut ~sep:"2." s with
-        | Some ("", _) -> `v2
-        | _ -> Fmt.failwith "opam: invalid version %s" s )
-  in
-  match v with
-  | Ok (v, (_, `Exited 0)) -> of_str v
-  | Ok (_, (_, s)) -> Fmt.failwith "opam: %a" OS.Cmd.pp_status s
-  | Error (`Msg e) -> Fmt.failwith "opam: %s" e
+module Version = struct
+  type t = V1_2_2 | V2
+
+  let pp fs = function
+    | V1_2_2 -> Format.fprintf fs "v1.2.2"
+    | V2 -> Format.fprintf fs "v2"
 
-let version = lazy (opam_version ())
+  let equal v1 v2 =
+    match (v1, v2) with V1_2_2, V1_2_2 | V2, V2 -> true | _ -> false
+
+  let of_string v =
+    if Bos_setup.String.is_prefix v ~affix:"1." then
+      if String.equal v "1.2.2" then Ok V1_2_2
+      else R.error_msgf "unsupported opam version: %S" v
+    else
+      match String.cut ~sep:"2." v with
+      | Some ("", _) -> Ok V2
+      | _ -> R.error_msgf "unsupported opam version: %S" v
+
+  let cli () =
+    match
+      OS.Cmd.run_out Cmd.(cmd % "--version") |> OS.Cmd.out_string ~trim:true
+    with
+    | Ok (v, (_, `Exited 0)) ->
+        if Bos_setup.String.is_prefix v ~affix:"1." then
```

Comment:
This should probably checked after parsing:
```ocaml
of_string v >>= fun v ->
if Opam.Version.equal v V1_2_2 then
  Logs.warn ... ;
Ok v
```

Reply:
done